kong/plugins/aws-lambda/handler.lua

```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
   AWSLambdaHandler.super.new(self, "aws-lambda")
 end
 
-local function retrieve_parameters()
-  ngx_req_read_body()
+local function retrieve_parameters(content_type, conf)
   local body_parameters, err
-  local content_type = ngx_req_get_headers()[CONTENT_TYPE]
-  if content_type and string_find(content_type:lower(), "multipart/form-data", nil, true) then
+  if content_type and string_find(content_type, "multipart/form-data", nil, true) then
     body_parameters = Multipart(ngx_req_get_body_data(), content_type):get_all()
-  elseif content_type and string_find(content_type:lower(), "application/json", nil, true) then
+  elseif content_type and string_find(content_type, "application/json", nil, true)
+         and not conf.request_body then
     body_parameters, err = cjson.decode(ngx_req_get_body_data())
     if err then
       body_parameters = {}
     end
-  else
+  elseif content_type and string_find(content_type, "application/x-www-form-urlencoded", nil, true) then
     body_parameters = public_utils.get_post_args()
+  else
+    body_parameters = {}
   end
 
   return utils.table_merge(ngx_req_get_uri_args(), body_parameters)
 end
 
+local function retrieve_body(content_type)
+  local err, body
+  local raw_body = ngx_req_get_body_data()
+  if raw_body == nil then return nil end
+
+  if content_type and string_find(content_type, "application/json", nil, true) then
+    body, err = cjson.decode(raw_body)
+    if err then
+      return nil
+    end
+  elseif content_type and ( string_find(content_type, "application/xml", nil, true)
+    or  string_find(content_type, "application/soap+xml", nil, true)
+    or  string_find(content_type, "text/", nil, true) )
+  then
+    body = raw_body
+  else
+    body = ngx_encode_base64(raw_body)
+  end
+  return body
+end
+
 function AWSLambdaHandler:access(conf)
   AWSLambdaHandler.super.access(self)
 
-  local bodyJson = cjson.encode(retrieve_parameters())
+  local content_type = ngx.var.http_content_type
+  if content_type then
+     content_type = content_type:lower()
+  end
+
+  ngx_req_read_body()
 
-  local host = string.format("lambda.%s.amazonaws.com", conf.aws_region)
-  local path = string.format("/2015-03-31/functions/%s/invocations",
+  local body_parameters = retrieve_parameters(content_type, conf)
```

Comment:
What about this? It would be still not breaking change and it still totally makes sense for me.
`  if conf.forward_request_body or conf.forward_http_headers 
     or conf.forward_http_method or conf.forward_request_uri then
     upstream_body = {
       request_body_args = not conf.forward_request_body and retrieve_parameters(content_type)
                           or ngx_req_get_uri_args(), 
       request_body = conf.forward_request_body and retrieve_body(content_type) or nil,
       request_http_headers = conf.forward_http_headers and ngx_req_get_headers() or nil,
       request_http_method = conf.forward_http_method and var.request_method or nil,
       request_uri = conf.forward_request_uri and var.request_uri or nil
     }
  else
     upstream_body = retrieve_parameters(content_type)
  end
`